### PR TITLE
EpicsSignal.set() timeout should default to EpicsSignal.timeout

### LINKED ISSUE #925: Why default `timeout` kwarg in EpicsSignal `.set()` not using signal's `.timeout`?
Why the `timeout` kwarg in EpicsSignal `.set()` method does not default to the `.timeout` value?  Looking here:  https://github.com/bluesky/ophyd/blob/3dddf111946c1eb7200361d2158445db2acfb76d/ophyd/signal.py#L1719

NOTE: The PV in use here is the `.STUP` (status update) field of an EPICS motor record.  Set it to 1, "ON" and the record support will set it back to 0 "OFF" as the motor status is updated.  Thus, the ophyd `.set` operation will not succeed since the PV will never reach the "ON" value.

This `.set()` will default the `timeout` kwarg to `None`.  Since I've set the default timeout for the signal to be 10 s, I was expecting `s.timeout` to be 10 (and it is).  

```
(bluesky_2020_9) prjemian@poof /tmp $ ipython
Python 3.8.2 (default, Mar 26 2020, 15:53:00) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.16.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from ophyd.signal import EpicsSignalBase                                                                                  

In [2]: EpicsSignalBase.set_default_timeout(timeout=10, connection_timeout=5)                                                     

In [3]: from ophyd import EpicsSignal                                                                                             

In [4]: s = EpicsSignal("sky:m1.STUP", name='s', string=True)                                                                     

In [5]: s.timeout                                                                                                                 
Out[5]: 10
```

Now, try `st = s.set("ON")` and wait 15 seconds (5 seconds beyond timeout) before printing the status.  Expecting it be done.

```
import time
st = s.set("ON")
time.sleep(15)
print(st)
```

Expecting `done`, but no.

```
In [6]: import time 
   ...: st = s.set("ON") 
   ...: time.sleep(15) 
   ...: print(st) 
   ...:                                                                                                                           
Status(obj=EpicsSignal(read_pv='sky:m1.STUP', name='s', value='OFF', timestamp=1603771932.302131, auto_monitor=False, string=True, write_pv='sky:m1.STUP', limits=False, put_complete=False), done=False, success=False)
```

---

Repeat but use a `timeout=10` kwarg with `.set()` in a new ipython session:

```
In [1]: from ophyd import EpicsSignal 
   ...: import time 
   ...: s = EpicsSignal("sky:m1.STUP", name='s', string=True) 
   ...: st = s.set("ON", timeout=10) 
   ...: time.sleep(15) 
   ...: print(st) 
   ...:                                                                                                                           
set_and_wait(value=ON, timeout=10, atol=None, rtol=None)
Status(obj=EpicsSignal(read_pv='sky:m1.STUP', name='s', value='OFF', timestamp=1603772109.341146, auto_monitor=False, string=True, write_pv='sky:m1.STUP', limits=False, put_complete=False), done=True, success=False)

```

That shows the expected sequence.  Signal is set, status timeout reached and reported, final print shows status is done but not successful.

_Originally posted by @prjemian in https://github.com/bluesky/ophyd/pull/901#issuecomment-716968738_

### PR DESCRIPTION
**Pull request by prjemian**
Changes to fix #925.  Expected the `EpicsSignal.set()` method to use the signal's `timeout` attribute.  This PR will make that happen.

### CODE REVIEW

**Comment by prjemian** [comment]
See #925 for details.

**Comment by prjemian** [comment]
Also, it would aid the diagnostics **immensely** if these reports included the signal's name: https://github.com/bluesky/ophyd/blob/3dddf111946c1eb7200361d2158445db2acfb76d/ophyd/signal.py#L284-L303
Since they are inside the `.set()` method, there is easy access to that information.

File: ophyd/signal.py [hunk 1]
```diff
@@ -1711,6 +1711,9 @@ def set(self, value, *, timeout=None, settle_time=None):
         --------
         Signal.set
         '''
+        if timeout is None:
+            timeout = self.timeout
```
Comment (tacaswell) [inline ophyd/signal.py:1715]: In `put` we use `self.write_timeout` instead of `self.timeout`, we should probably be consistent with that here.

The other "fix" may be to change the default value of timeout to be `timeout=DEFAULT_WRITE_TIMEOUT` like put (as I assume it eventually falls through to that).

File: ophyd/signal.py [hunk 2]
```diff
@@ -287,20 +287,20 @@ def set_thread():
             except TimeoutError:
                 success = False
                 self.log.warning(
-                    'set_and_wait(value=%s, timeout=%s, atol=%s, rtol=%s)',
-                    value, timeout, self.tolerance, self.rtolerance
+                    'set_and_wait(name=%s, value=%s, timeout=%s, atol=%s, rtol=%s)',
```
Comment (tacaswell) [inline ophyd/signal.py:290]: :+1:

Comment (tacaswell) [reply ophyd/signal.py:290]: one concern is that I don't think we can actually pass `name` to `set_and_wait`, but I like the idea.

File: ophyd/signal.py [hunk 3]
```diff
@@ -1711,6 +1711,9 @@ def set(self, value, *, timeout=None, settle_time=None):
         --------
         Signal.set
         '''
+        if timeout is None:
```
Comment (tacaswell) [inline ophyd/signal.py:1714]: This sentenel should now check for `DEFAULT_WRITE_TIMEOUT`?

File: ophyd/signal.py (hunk 2 above)
Comment (prjemian) [reply ophyd/signal.py:290]: In _this particular_ case, `self.name` is accessible since the log comes from an inner (?what's the right name for this type of thingy?) function and has access to all its parent's attributes.

File: ophyd/signal.py (hunk 3 above)
Comment (prjemian) [reply ophyd/signal.py:1714]: Ok, now I understand this usage pattern.  `DEFAULT_WRITE_TIMEOUT` is a defined object, different from `None` but meant to show this value is not defined yet, right?

File: ophyd/signal.py (hunk 1 above)
Comment (prjemian) [reply ophyd/signal.py:1715]: When I check my test case, `self.write_timeout` is `None` while `self.timeout` is as defined (by me) via `EpicsSignalBase.set_default_timeout(timeout=10, connection_timeout=5)`.

So we need a second test for `None`ness.

Comment (prjemian) [reply ophyd/signal.py:1715]:
```
In [6]: s.write_timeout is None                                                                                                                    
Out[6]: True

In [7]: s.timeout                                                                                                                                  
Out[7]: 10
```

File: ophyd/signal.py (hunk 3 above)
Comment (tacaswell) [reply ophyd/signal.py:1714]: Yeah, we need to sort out "did the user not pass _anything_ in which case we do our version of the default, otherwise just pass through to the underlying level" so we can thread `timeout=None` through to mean "no timeout".

Comment (prjemian) [reply ophyd/signal.py]: see c94f3ae 

This preserves the intent if user passed `timeout=None` meaning _do not timeout_.  It tries to use `self.write_timeout` but if _that_ was `None`, uses `self.timeout`.  Oy!

File: ophyd/signal.py (hunk 2 above)
Comment (tacaswell) [reply ophyd/signal.py:290]: Sorry, I meant that `set_and_wait(name='xyz', ..)` won't work (but it does look nice in the log!)

Would `'%s: set_and_wait(...)'` be better?

File: ophyd/signal.py [hunk 4]
```diff
@@ -1711,6 +1711,11 @@ def set(self, value, *, timeout=None, settle_time=None):
         --------
         Signal.set
         '''
+        if timeout is DEFAULT_WRITE_TIMEOUT:
+            timeout = self.write_timeout
+            if timeout is None:
```
Comment (tacaswell) [inline ophyd/signal.py:1716]: I do not think we want to fallback to the read time out here given that they are independently configurable https://github.com/bluesky/ophyd/blob/33bfc595ac2862aede96f023296f564832144551/ophyd/signal.py#L679-L699

and in the init we have a long comment differentiating the two https://github.com/bluesky/ophyd/blob/33bfc595ac2862aede96f023296f564832144551/ophyd/signal.py#L757-L765

Comment (tacaswell) [reply ophyd/signal.py:1716]: We also don't do this fallback in `put` https://github.com/bluesky/ophyd/blob/33bfc595ac2862aede96f023296f564832144551/ophyd/signal.py#L1661-L1665

Comment (prjemian) [reply ophyd/signal.py:1716]: My original version (setting `timeout` from `self.timeout`) was my best understanding of what to do,  Then I learned about `DEFAULT_WRITE_TIMEOUT`.  Now I'm puzzled what to do.  I thought the default timeout that was set in `EpicsSignalBase.set_default_timeout(timeout=10, connection_timeout=5)` would apply in this situation.

Can you suggest the code to use?

Comment (prjemian) [reply ophyd/signal.py:1716]: Perhaps this is the one to use as a fallback? https://github.com/bluesky/ophyd/blob/33bfc595ac2862aede96f023296f564832144551/ophyd/signal.py#L866

Comment (prjemian) [reply ophyd/signal.py:1716]: The reason I'm puzzled is that `self.write_timeout` is `None` in my test case and has not been affected by the `set_default_timeout()` method.  I assumed it would be.  Perhaps fallback (when `write_timeout` is unconfigured) to `__default_timeout ` fits best what I assumed would happen.

**Comment by tacaswell** [comment]
I think this history here is:

 - we did the big "add global (read) timeout settings" change (via https://github.com/bluesky/ophyd/pull/788)
 - some time later when we tried to use this in anger on the floor we discovered we had only done the read (not write side) and added config for that too (via https://github.com/bluesky/ophyd/pull/804)
 - as part of #804 we forgot to thread the configuration of the default through the class level `set_default_timeout` classmethod

The naming on this is also confusing because in the read methods everything is called `timeout`, but in the write methods the local variables and function parameters are still called `timeout`, but the internal state is `write_timeout`.  In the init we have both `timeout` and `write_timeout` as kwargs.

**Comment by prjemian** [comment]
Following #804, is it reasonable to support configuration of the default write timeout via `set_default_timeout()`?

**Comment by tacaswell** [comment]
Yes

**Comment by prjemian** [comment]
Good, I'll focus on that next, then we can circle back to this code and _make it right_, too.

**Comment by prjemian** [comment]
Alternatively, the existing` set_default_timeout()` method and a proposed `set_default_auto_monitor()` method (1) could be combined into a single `set_defaults()` method and thus remove duplicated code and comments.

1. #927, from https://github.com/BCDA-APS/apstools/issues/450#issuecomment-719640338

File: ophyd/signal.py [hunk 5]
```diff
@@ -858,15 +869,27 @@ def set_default_timeout(cls, *, timeout=2.0, connection_timeout=1.0):
...
+        cls.__default_write_timeout = write_timeout
 
-    # TODO Is there a good reason to prohibit setting these three timeout
-    # properties?
+        # TODO Is there a good reason to prohibit setting these three timeout
+        # properties?
+
+    @classmethod
+    def set_default_timeout(cls, *, **kwargs):
+        warnings.warn('Use set_defaults() instead.  Command ignored.')
```
Comment (danielballan) [inline ophyd/signal.py:892]: I would prefer to pass through to `cls.set_defaults(**kwargs)` here so that this change is backward compatible. We can warn about _future_ removal.

**Review by danielballan — COMMENTED** [review]
This looks ready to me, minus one point about backward-compatibility marked inline.

File: ophyd/signal.py [hunk 6]
```diff
@@ -858,15 +869,27 @@ def set_default_timeout(cls, *, timeout=2.0, connection_timeout=1.0):
...
+        # being too long and delaying the report of a failure.
+        cls.__default_write_timeout = write_timeout
 
-    # TODO Is there a good reason to prohibit setting these three timeout
-    # properties?
+        # TODO Is there a good reason to prohibit setting these three timeout
+        # properties?
+
+    @classmethod
+    def set_default_timeout(cls, *, **kwargs):
```
Comment (danielballan) [inline ophyd/signal.py:891]: This is a syntax error. You can either do `(cls, *args, **kwargs)` or `(cls, **kwargs)`. The latter means, "Only kwargs are accepted," which seems like your intent and the correct thing to do here, since `set_default_timeout` only ever accepted keyword arguments.

```suggestion
    def set_default_timeout(cls, **kwargs):
```

Comment (prjemian) [reply ophyd/signal.py:891]: Good catch.  One fine day, I will learn this pattern.

File: ophyd/signal.py (hunk 5 above)
Comment (prjemian) [reply ophyd/signal.py]: Kinder, gentler, done

**Review by danielballan — APPROVED** [review]
